bridges/modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -67,148 +100,276 @@ pub const HARD_MESSAGE_SIZE_LIMIT: u32 = 32 * 1024;
-	#[pallet::getter(fn delivery_fee_factor)]
-	pub type DeliveryFeeFactor<T: Config<I>, I: 'static = ()> =
-		StorageValue<_, FixedU128, ValueQuery, InitialFactor>;
+	pub type Bridges<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Blake2_128Concat, BridgeIdOf<T, I>, BridgeState, OptionQuery>;
 
 	impl<T: Config<I>, I: 'static> Pallet<T, I> {
-		/// Called when new message is sent (queued to local outbound XCM queue) over the bridge.
-		pub(crate) fn on_message_sent_to_bridge(message_size: u32) {
-			// if outbound channel is not congested, do nothing
-			if !T::LocalXcmChannelManager::is_congested(&T::SiblingBridgeHubLocation::get()) {
+		/// Called when new message is sent to the `dest` (queued to local outbound XCM queue).
+		pub(crate) fn on_message_sent_to(message_size: u32, dest: Location) {
+			let Some(bridge_id) = T::BridgeIdResolver::resolve_for_dest(&dest) else {
+				// not supported bridge id, so do nothing
 				return
-			}
-
-			// ok - we need to increase the fee factor, let's do that
-			let message_size_factor = FixedU128::from_u32(message_size.saturating_div(1024))
-				.saturating_mul(MESSAGE_SIZE_FEE_BASE);
-			let total_factor = EXPONENTIAL_FEE_BASE.saturating_add(message_size_factor);
-			DeliveryFeeFactor::<T, I>::mutate(|f| {
-				let previous_factor = *f;
-				*f = f.saturating_mul(total_factor);
+			};
+
+			// handle congestion and fee factor (if detected)
+			let increased =
+				Bridges::<T, I>::mutate_exists(&bridge_id, |bridge_state| match bridge_state {
+					Some(ref mut bridge_state) if bridge_state.is_congested => {
+						// found congested bridge
+						// ok - we need to increase the fee factor, let's do that
+						let message_size_factor =
+							FixedU128::from_u32(message_size.saturating_div(1024))
+								.saturating_mul(MESSAGE_SIZE_FEE_BASE);
+						let total_factor = EXPONENTIAL_FEE_BASE.saturating_add(message_size_factor);
+
+						let previous_factor = bridge_state.delivery_fee_factor;
+						bridge_state.delivery_fee_factor =
+							bridge_state.delivery_fee_factor.saturating_mul(total_factor);
+
+						Some((previous_factor, bridge_state.delivery_fee_factor))
+					},
+					_ => {
+						// not congested, do nothing
+						None
+					},
+				});
+			if let Some((previous_factor, new_factor)) = increased {
 				log::info!(
-					target: LOG_TARGET,
-					"Bridge channel is congested. Increased fee factor from {} to {}",
-					previous_factor,
-					f,
+						target: LOG_TARGET,
+						"Bridge channel with id {:?} is congested. Increased fee factor from {} to {} for {:?}",
+						bridge_id,
+						previous_factor,
+						new_factor,
+						dest
 				);
-				Self::deposit_event(Event::DeliveryFeeFactorIncreased { new_value: *f });
-				*f
+				Self::deposit_event(Event::DeliveryFeeFactorIncreased {
+					previous_value: previous_factor,
+					new_value: new_factor,
+					bridge_id,
+					dest,
+				});
+			}
+		}
+
+		/// Calculates dynamic fees for a given asset based on the bridge state.
```

Comment:
Aren't the fees paid in only one asset?

Reply:
good point, I cleaned it a little bit: https://github.com/paritytech/polkadot-sdk/pull/6231/commits/b8d274222731aef412a6c7b290d36838a239c752